### ISSUE TITLE
Provide more usable columns environment

### ISSUE THREAD

**Issue by josephwright** [josephwright/beamer#262]
Originally reported by: **svenkoehler (Bitbucket: [svenkoehler](https://bitbucket.org/svenkoehler), GitHub: [svenkoehler](https://github.com/svenkoehler))**

----------------------------------------

For the last presentation I made, the columns environment was a complete nightmare. What I want is a predictable width for some of the columns and for the sum of columns widths (including the space between columns). In the following I discuss the various issues with the columns environment and describe two solutions.

Horizontal alignment and column widths:
Something like having two columns with 0.5\textwidth each will result in two columns with a width of 0.5\textwidth and some space in between them. While that is totally what one might expect, it of course leads to the fact that the text in the left column of that slide will be positioned much more to the left than on other slides that don't use columns. That looks bad, and breaks the transition between slides with and without columns.
The onlytextwidth option is available, but to the best of my knowledge, it basically scales down the width of all columns proportionally, such that the sum of all widths matches \textwidth. Now that gives me what I want regarding the aspects total width and positioning of the text, but of course the actual widths of the columns become kind of unpredictable. What, if I want the right column to be exactly 3cm in width? Having one column of width \textwidth-3cm and another with width 3cm doesn't do the job if the option onlytextwidth is used. And without onlytextwidth, the total with is too large.
An example for a column that is exactly 3cm in width is the case, that the column contains a tikz picture, that is exactly 3cm in width (the bounding box of tikz images can be easily controlled). This ensures positioning of text and graphics consistently over different slides.

Solutions:
The easiest solution is that the user has to do the math. I could have one column of width \textwidth-\colsep-3cm and another one with width \3cm. I wouldn't need to use the onlytextwidth option, and would be a happy beamer user. However, the \colsep macro doesn't exist (or means something else). To the best of my knowledge, such a macro (actually it should be a length, yes?) should be introduced. It should be easily accessible, and the documentation should mention it.

The most convenient solution would be to have a special value like "!" such that one could define one or more columns width a fixed width, and the column with width ! would take up the rest of the space, such that the sum of all widths (including the space inbetween columns) would equal some desired total width value (e.g. \textwidth by default).

----------------------------------------
- Bitbucket: https://bitbucket.org/rivanvx/beamer/issue/262

**Comment by josephwright** [comment]
*Original comment by* **Joseph Wright (Bitbucket: [josephwright](https://bitbucket.org/josephwright), GitHub: [josephwright](https://github.com/josephwright))**:

----------------------------------------

The code for `onlytextwidth` doesn't do any scaling at all: all that happens is that it does not add any margins. As such, I think that this option already covers the case you seem to want. For example
````
\documentclass{beamer}
\begin{document}
\begin{frame}
  ABC
\end{frame}
\begin{frame}
\begin{columns}[onlytextwidth]
  \begin{column}{0.3\textwidth}
    ABC
  \end{column}
  \begin{column}{0.3\textwidth}
    123
  \end{column}
  \begin{column}{0.4\textwidth}
    XYZ
  \end{column}
\end{columns}
\end{frame}
\end{document}
````
seems to have the correct horizontal placement, to me.

Making a change to the column code could well be risky: I have enough problems with bits of the code that are apparently much less complex! I tend to try to stick to what Till decided on if it looks deliberate as while I might question some of the decisions they've been in use 'in the wild' for quite some time.

As such, this is either 'INVALID' (as the behaviour is already available) or 'WONTFIX' (as I'm not sure I want to try this). I'll wait for some feedback before closing.

**Comment by josephwright** [comment]
*Original comment by* **Joseph Wright (Bitbucket: [josephwright](https://bitbucket.org/josephwright), GitHub: [josephwright](https://github.com/josephwright))**:

----------------------------------------

Also worth noting is that the code doesn't add any space between columns, which are implemented as `minipage`s.

**Comment by josephwright** [comment]
*Original comment by* **svenkoehler (Bitbucket: [svenkoehler](https://bitbucket.org/svenkoehler), GitHub: [svenkoehler](https://github.com/svenkoehler))**:

----------------------------------------

Thanks for correcting my misconception of the onlytextwidth option.
However, if that option is not used, there seems to be some space between columns. In that case, there seems no easy way of having one column of say 3cm, and one columns that takes the rest. Among my suggestions for making this possible, there is a very simple and non-intrusive solution: provide a length macro, that holds how much space is inserted between columns. Then one can one column 3cm and one column of \textwidth-\foo-3cm, where \foo would be the space. Using that kind of arithmetic should already be covered by the calc package, right?

**Comment by josephwright** [comment]
*Original comment by* **Joseph Wright (Bitbucket: [josephwright](https://bitbucket.org/josephwright), GitHub: [josephwright](https://github.com/josephwright))**:

----------------------------------------

There isn't any space inserted between columns: they are just vboxes set next to one another

**Comment by josephwright** [comment]
*Original comment by* **svenkoehler (Bitbucket: [svenkoehler](https://bitbucket.org/svenkoehler), GitHub: [svenkoehler](https://github.com/svenkoehler))**:

----------------------------------------

You're wrong. There is space between columns. But it's not fixed, as I thought. Beamer will try to have the same space between the borders of the page and the first/last column and inbetween the columns. I guess, there are a bunch of \hfill statements between the minipages?

Only if you specify the totalwidth option, columns will suddenly be aligned to the left border of the slides textarea (not the page) and space between the columns will only be added if the sum of the column widths is less than the totalwidth.

The documentation doesn't tell you anything about the alignment of the columns or the space inbetween them. In particular, the side effect of totalwidth (which changes the columns to be left aligned within the textarea of the slide rather than center aligned on the page).  If the documentation had a face, I would punch it.

Here's code that illustrates what I mean:

```
#!latex
\documentclass{beamer}
\usepackage{fix-cm}
\usepackage{tikz}
\begin{document}
\begin{frame}[t]
\the\textwidth
\begin{columns}[t,totalwidth=7cm]
\column{2cm}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\column{2cm}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\column{2cm}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\end{columns}
\begin{columns}[t]
\column{3.5cm}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\column{3.5cm}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\column{3.5cm}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\end{columns}
\end{frame}
\end{document}
```

**Comment by josephwright** [comment]
*Original comment by* **svenkoehler (Bitbucket: [svenkoehler](https://bitbucket.org/svenkoehler), GitHub: [svenkoehler](https://github.com/svenkoehler))**:

----------------------------------------

I should add, that if you have two columns of 0.5\textwidth and don't specify totalwidth, then you automatically get space between them. Also, the left column will start significantly left of what normal slides start at. Not only does it look bad, it is also very confusing.

```
#!latex
\begin{columns}[t]
\column{0.5\textwidth}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\column{0.5\textwidth}
	\the\textwidth\\
	\begin{tikzpicture}\fill (0,0) rectangle (\textwidth,1);\end{tikzpicture}
\end{columns}
```

I shall open a new issue about the documentation issues.

**Comment by josephwright** [comment]
*Original comment by* **Joseph Wright (Bitbucket: [josephwright](https://bitbucket.org/josephwright), GitHub: [josephwright](https://github.com/josephwright))**:

----------------------------------------

OK, I'll reopen and trace the code though again

**Comment by josephwright** [comment]
*Original comment by* **svenkoehler (Bitbucket: [svenkoehler](https://bitbucket.org/svenkoehler), GitHub: [svenkoehler](https://github.com/svenkoehler))**:

----------------------------------------

Thanks for reopening, but I'm not sure my original problem still exists.
Now that I know how things work, it's quite easy to have columns of defined width with a defined space between them. For example, you can have one column of 3cm, and one of \textwidth-3cm-1cm. With totalwidth=\textwidth, that gives you 1cm space between the two columns.

Personally, I don't like that the different semantics regarding the horizontal placement of the columns if the totalwidth option is not present. Why would the columns environment completely ignore the left and right margin defines by the theme or the author of the presentation? I guess, Till made this choice? I will always go with totalwidth=\textwidth from now on.

If you want to make this issue about fixing the documentation, i'm fine with it. I can also happy to open a new issue. If you point me at the source of the documentation, I'm happy to contribute a patch for the documentation of the columns environment that properly explains how columns are placed on the slide.

**Comment by josephwright** [comment]
*Original comment by* **Joseph Wright (Bitbucket: [josephwright](https://bitbucket.org/josephwright), GitHub: [josephwright](https://github.com/josephwright))**:

----------------------------------------

Yes, this decision is Till's: I've checked the history to be sure. I can't say why he did it, but will look to document.

**Comment by josephwright** [comment]
*Original comment by* **Joseph Wright (Bitbucket: [josephwright](https://bitbucket.org/josephwright), GitHub: [josephwright](https://github.com/josephwright))**:

----------------------------------------

Add a note to explain about margins and column width (closes #262)

Not sure how clear this is, but hopefully covers at least some of it!
